### ISSUE TITLE
Over-zealous "error non-sargable : Performing functions on filter clauses or join predicates can cause performance problems." rule?

### ISSUE THREAD

**Issue by matt-s-wise** [tsqllint/tsqllint#184]
Consider the following query:
```
	SELECT d.ID AS DebtorId, addr.Address
	  FROM dbo.Debtor d
	 INNER JOIN (
		SELECT DebtorId, Address, ROW_NUMBER() OVER (PARTITION BY DebtorId ORDER BY ID DESC) AS rownum
		  FROM dbo.Address
	) addr ON d.ID = addr.DebtorID
	WHERE addr.rownum = 1;
```
It reports the non-sargable error because of the use of the ROW_NUMBER() function that is then filtered in the outer SELECT.
Unless I'm missing something, this should be a perfectly valid scenario; it can be rewritten as a CROSS APPLY to eliminate the warning but from what I've read, whether a JOIN or APPLY performs better than the other depends on the shape of the data.
Another way of silencing the warning is to extract the JOIN subquery to a CTE - but the execution plan is identical, so the error should either be reported for both or for neither.

**Comment by nathan-boyd** [comment]
reproduced agreed.

**Comment by nathan-boyd** [comment]
resolved in #185
